drivers/network/netio/netio.c

```diff
@@ -0,0 +1,1453 @@
+                            /* Windows specific. This sets the mask for callback functions: */
+                        case SO_WSK_EVENT_CALLBACK:
+                            if (InputBuffer == NULL)
+                            {
+                                DPRINT1("WskControlSocket: Need an InputBuffer for this operation\n");
+                                status = STATUS_INVALID_PARAMETER;
+                                break;
+                            }
+                            if (InputSize < sizeof(WSK_EVENT_CALLBACK_CONTROL))
+                            {
+                                DPRINT1("WskControlSocket: InputBuffer too small for this operation\n");
+                                status = STATUS_INVALID_PARAMETER;
+                                break;
+                            }
+                            WSK_EVENT_CALLBACK_CONTROL *c = (WSK_EVENT_CALLBACK_CONTROL *)InputBuffer;
+
+                            if (((s->CallbackMask & WSK_EVENT_ACCEPT) == 0) &&
+                                ((c->EventMask & WSK_EVENT_ACCEPT) == WSK_EVENT_ACCEPT))
+                            {
+                                s->CallbackMask = c->EventMask;
+                                QueueListening(s);
+                            }
+                            else
+                            {
+                                s->CallbackMask = c->EventMask;
+                            }
+
+                            status = STATUS_SUCCESS;
+                            break;
+
+                        default:
+                            DPRINT1("WskControlSocket: ControlCode %d Not implemented\n", ControlCode);
+                    }
+                    break;
+                default:
+                    DPRINT1("WskControlSocket: Level %d Not implemented\n", Level);
+            }
+            break;
+
+        case WskGetOption:
+        case WskIoctl:
+        default:
+            DPRINT1("WskControlSocket: Option %d Not implemented\n", RequestType);
+    }
+
+err_out:
+    Irp->IoStatus.Status = status;
+    IoCompleteRequest(Irp, IO_NETWORK_INCREMENT);
+
+    return status;
+}
+
+static NTSTATUS WSKAPI
+WskCloseSocket(_In_ PWSK_SOCKET Socket, _Inout_ PIRP Irp)
+{
+    NTSTATUS status = STATUS_SUCCESS;
+    PWSK_SOCKET_INTERNAL s = (PWSK_SOCKET_INTERNAL)Socket;
+
+    IoSetNextIrpStackLocation(Irp);
+
+    /* There might be a reference from (for example) a pending
+     * receive. Shutdown the socket here explicitly. We expect
+     * all pending I/O operations to be cancelled, then.
+     */
+
+    SocketShutdown(s);
+    SocketPut(s);
+
+    Irp->IoStatus.Status = status;
+    IoCompleteRequest(Irp, IO_NETWORK_INCREMENT);
+
+    return STATUS_PENDING;
+}
+
+static PTRANSPORT_ADDRESS
+TdiTransportAddressFromSocketAddress(_In_ PSOCKADDR SocketAddress)
+{
+    PTRANSPORT_ADDRESS ta;
+
+    ta = ExAllocatePoolWithTag(NonPagedPool, sizeof(*ta) + sizeof(SocketAddress->sa_data), TAG_NETIO);
```

Comment:
if I understand you correctly, you can consider this way:
```suggestion
    ULONG Size;
    
    Size = FIELD_OFFSET(TRANSPORT_ADDRESS, Address) +
           FIELD_OFFSET(TA_ADDRESS, Address[sizeof(SocketAddress->sa_data)]);
           
    ta = ExAllocatePoolUninitialized(NonPagedPool, Size, TAG_NETIO);
```

Reply:
Hi, I am not sure ... I think sizeof() is the better way to express this. I will leave this open for now.

Reply:
Ok I applied it, thanks